src/parse.rs

```diff
@@ -168,34 +170,125 @@ impl<'a> Bytes<'a> {
         Ok(())
     }
 
-    pub fn advance_byte(&mut self) -> Result<()> {
-        if self.peek_byte_or_eof()? == b'\n' {
+    pub fn advance_char(&mut self) -> Result<usize> {
+        self.next().map(char::len_utf8)
+    }
+
+    pub fn next(&mut self) -> Result<char> {
+        let c = self.peek()?;
+        if c == '\n' {
             self.cursor.line += 1;
             self.cursor.col = 1;
         } else {
             self.cursor.col += 1;
         }
-        self.string = &self.string[1..];
 
-        Ok(())
+        self.src = &self.src[c.len_utf8()..];
+
+        Ok(c)
     }
 
-    pub fn advance_char(&mut self) -> Result<usize> {
-        let c = self.peek_char_or_eof()?;
-        if c == '\n' {
-            self.cursor.line += 1;
-            self.cursor.col = 1;
+    pub fn peek(&self) -> Result<char> {
+        self.src.chars().next().ok_or(Error::Eof)
+    }
+
+    pub fn bytes(&self) -> &'a [u8] {
```

Comment:
`escaped_string`, and the start of both `string` and `integer`

Reply:
I think what I was wondering about is if there are any cases left where we need to check bytes instead of chars for functionality. Do you think code that only relied on chars would be cleaner?